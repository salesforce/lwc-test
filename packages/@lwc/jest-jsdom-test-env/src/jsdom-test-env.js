const JSDOMEnvironment = require('jest-environment-jsdom').TestEnvironment;

/**
 * Custom Jest environment extending JSDOM.
 */
class CustomJSDOMEnvironment extends JSDOMEnvironment {
    /**
     * Initializes the custom environment.
     *
     * @param {object} config - Jest configuration.
     * @param {object} context - Test context, including file path.
     */
    constructor(config, context) {
        super(config, context);
        this.testPath = context.testPath;
    }

    /**
     * Sets up the environment, including the test file path.
     */
    async setup() {
        await super.setup();
        this.global.testFilePath = this.testPath;
    }

    /**
     * Cleans up the environment.
     */
    async teardown() {
        await super.teardown();
    }
}

module.exports = CustomJSDOMEnvironment;
