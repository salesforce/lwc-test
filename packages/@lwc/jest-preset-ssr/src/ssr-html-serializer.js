/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
const HTML_NAMESPACE = 'http://www.w3.org/1999/xhtml';

// Void elements are elements that self-close even without an explicit solidus (slash),
// e.g. `</tagName>` or `<tagName />`. For instance, `<meta>` closes on its own; no need for a slash.
// These only come from HTML; there are no void elements in the SVG or MathML namespaces.
// See: https://html.spec.whatwg.org/multipage/syntax.html#syntax-tags
const VOID_ELEMENTS_SET = new Set([
    'area',
    'base',
    'br',
    'col',
    'embed',
    'hr',
    'img',
    'input',
    'link',
    'meta',
    'source',
    'track',
    'wbr',
]);

function isVoidElement(name, namespace) {
    return namespace === HTML_NAMESPACE && VOID_ELEMENTS_SET.has(name.toLowerCase());
}

// Note: Original code lives here: https://github.com/salesforce/lwc/blob/master/packages/%40lwc/engine-server/src/__tests__/fixtures.spec.ts#L54-L117
/**
 * Naive HTML fragment formatter.
 *
 * This is a replacement for Prettier HTML formatting. Prettier formatting is too aggressive for
 * fixture testing. It not only indent the HTML code but also fixes HTML issues. For testing we want
 * to make sure that the fixture file is as close as possible to what the engine produces.
 *
 * @param src the original HTML fragment.
 * @returns the formatter HTML fragment.
 */
function formatHTML(src) {
    let res = '';
    let pos = 0;
    let start = pos;

    let depth = 0;

    const getPadding = () => {
        return '  '.repeat(depth);
    };

    while (pos < src.length) {
        // Consume element tags and comments.
        if (src.charAt(pos) === '<') {
            const tagNameMatch = src.slice(pos).match(/(\w+)/);

            const isVoid = isVoidElement(tagNameMatch[0], HTML_NAMESPACE);
            const isClosing = src.charAt(pos + 1) === '/';
            const isComment =
                src.charAt(pos + 1) === '!' &&
                src.charAt(pos + 2) === '-' &&
                src.charAt(pos + 3) === '-';

            start = pos;
            while (src.charAt(pos++) !== '>') {
                // Keep advancing until the closing tag.
            }

            // Adjust current depth and print the element tag or comment.
            if (isClosing) {
                depth--;
            }

            res += getPadding() + src.slice(start, pos) + '\n';

            const isSelfClosing = src.charAt(pos - 2) === '/';
            if (!isClosing && !isSelfClosing && !isVoid && !isComment) {
                depth++;
            }
        }

        // Consume text content.
        start = pos;
        while (src.charAt(pos) !== '<' && pos < src.length) {
            pos++;
        }

        if (start !== pos) {
            res += getPadding() + src.slice(start, pos) + '\n';
        }
    }

    return res.trim();
}

module.exports = {
    test(val) {
        return typeof val === 'string' && val.length > 0;
    },
    serialize: formatHTML,
};