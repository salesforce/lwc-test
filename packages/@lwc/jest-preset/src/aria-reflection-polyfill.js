// Minimal polyfill for ARIA string reflection, e.g. aria-label -> ariaLabel.
// As of this writing, ARIA string reflection is supported in Chrome and Safari, but not Firefox or JSDOM.
// TODO: remove after JSDOM merges this PR: https://github.com/jsdom/jsdom/pull/3586

// ARIA Element properties that reflect to strings (not Element or FrozenArray<Element>)
// https://w3c.github.io/aria/#ARIAMixin
const ARIA_STRING_PROPS = [
    'role',
    'ariaAtomic',
    'ariaAutoComplete',
    'ariaBusy',
    'ariaChecked',
    'ariaColCount',
    'ariaColIndex',
    'ariaColIndexText',
    'ariaColSpan',
    'ariaCurrent',
    'ariaDescription',
    'ariaDisabled',
    'ariaExpanded',
    'ariaHasPopup',
    'ariaHidden',
    'ariaInvalid',
    'ariaKeyShortcuts',
    'ariaLabel',
    'ariaLevel',
    'ariaLive',
    'ariaModal',
    'ariaMultiLine',
    'ariaMultiSelectable',
    'ariaOrientation',
    'ariaPlaceholder',
    'ariaPosInSet',
    'ariaPressed',
    'ariaReadOnly',
    'ariaRequired',
    'ariaRoleDescription',
    'ariaRowCount',
    'ariaRowIndex',
    'ariaRowIndexText',
    'ariaRowSpan',
    'ariaSelected',
    'ariaSetSize',
    'ariaSort',
    'ariaValueMax',
    'ariaValueMin',
    'ariaValueNow',
    'ariaValueText',
];

for (const prop of ARIA_STRING_PROPS) {
    if (!(prop in Element.prototype)) {
        // sniff for support
        // e.g. `ariaPosInSet` -> `aria-posinset`
        const attribute = prop.replace(/[A-Z]/, (letter) => `-${letter}`).toLowerCase();
        Object.defineProperty(Element.prototype, prop, {
            get() {
                return this.getAttribute(attribute);
            },
            set(value) {
                // Per the spec, only null is treated as removing the attribute. However, Chromium/WebKit currently
                // differ from the spec and allow undefined as well. Here, we follow the standard.
                // See: https://github.com/w3c/aria/issues/1858
                if (value === null) {
                    this.removeAttribute(attribute);
                } else {
                    this.setAttribute(attribute, value);
                }
            },
            // These props in both WebKit and Chromium are configurable/enumerable. This allows overriding
            // (Try `Object.getOwnPropertyDescriptor(Element.prototype, 'ariaLabel')` in either browser)
            configurable: true,
            enumerable: true,
        });
    }
}
