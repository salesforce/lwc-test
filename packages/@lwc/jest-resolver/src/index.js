/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
const fs = require('fs');
const { resolve, extname, join, dirname, basename } = require('path');

const EMPTY_CSS_MOCK = resolve(__dirname, '..', 'resources', 'emptyStyleMock.js');
const EMPTY_HTML_MOCK = resolve(__dirname, '..', 'resources', 'emptyHtmlMock.js');

const ALLOWLISTED_LWC_PACKAGES = {
    lwc: '@lwc/engine-dom',
    'wire-service': '@lwc/wire-service',
    'wire-service-jest-util': '@salesforce/wire-service-jest-util',
};

// This logic is somewhat the same in the compiler resolution system
// We should try to consolidate it at some point.
function isImplicitHTMLImport(importee, { basedir }) {
    const ext = extname(importee);
    const isHTML = ext === '.html';
    const fileName = basename(importee, '.html');
    const absPath = resolve(basedir, importee);
    const dir = dirname(absPath);
    const jsFile = join(dir, fileName + '.js');
    const tsFile = join(dir, fileName + '.ts');

    return (
        // if is an HTML file
        isHTML &&
        // the html must not exist
        !fs.existsSync(absPath) &&
        // there must be a js/ts file with the same name in the same folder
        (fs.existsSync(jsFile) || fs.existsSync(tsFile))
    );
}

function isValidCSSImport(importee, { basedir }) {
    const ext = extname(importee);
    const isCSS = ext === '.css';
    let fileName = basename(importee, '.css');
    const isScoped = extname(fileName) === '.scoped';
    if (isScoped) {
        fileName = basename(fileName, '.scoped');
    }
    const absPath = resolve(basedir, importee);
    const dir = dirname(absPath);
    const jsFile = join(dir, fileName + '.js');
    const tsFile = join(dir, fileName + '.ts');

    return (
        // if it is a css file
        isCSS &&
        // the css file must exist
        fs.existsSync(absPath) &&
        // there must be a js/ts file with the same name in the same folder
        (fs.existsSync(jsFile) || fs.existsSync(tsFile))
    );
}

function getLwcPath(path, options) {
    if (path.endsWith('?scoped=true')) {
        // remove query param for scoped styles
        path = path.substring(0, path.length - 12);
    }
    // If is a special LWC package, resolve it from commonjs
    if (ALLOWLISTED_LWC_PACKAGES[path]) {
        return require.resolve(ALLOWLISTED_LWC_PACKAGES[path]);
    }

    // If is an implicit imported CSS just resolve it to an empty file
    if (extname(path) === '.css' && !isValidCSSImport(path, options)) {
        return EMPTY_CSS_MOCK;
    }

    // If is an implicit imported html (auto-binded from the compiler) return an empty file
    if (isImplicitHTMLImport(path, options)) {
        return EMPTY_HTML_MOCK;
    }

    return path;
}

module.exports = function (path, options) {
    return options.defaultResolver(getLwcPath(path, options), options);
};
