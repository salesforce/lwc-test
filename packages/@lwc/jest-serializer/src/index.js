/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
const PrettyFormat = require('pretty-format');
const DOMElement = PrettyFormat.plugins.DOMElement;

const cleanElementAttributes = require('./clean-element-attrs');

function test(obj) {
    if (typeof obj !== 'object' || obj === null) {
        return false;
    }

    const { nodeType } = obj;

    // According https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType
    return (
        nodeType === 1 || // element
        nodeType === 3 || // text
        nodeType === 8 // comment
    );
}

const { defineProperty } = Object;

function escapeHTML(str) {
    return str.replace(/</g, '&lt;').replace(/>/g, '&gt;');
}

function printText(text, config) {
    const contentColor = config.colors.content;
    return contentColor.open + escapeHTML(text) + contentColor.close;
}

function printChildren(children, config, indentation, depth, refs, printer) {
    return children
        .map(
            (child) =>
                config.spacingOuter +
                indentation +
                (typeof child === 'string'
                    ? printText(child, config)
                    : printer(child, config, indentation, depth, refs))
        )
        .join('');
}

function serialize(node, config, indentation, depth, refs, printer) {
    const isElement = node.nodeType === 1;
    if (isElement) {
        cleanElementAttributes(node);
    }

    const lightChildren = Array.prototype.slice.call(node.childNodes);
    defineProperty(node, 'childNodes', {
        get() {
            if (node.shadowRoot) {
                const textNode = document.createTextNode('#shadow-root(open)');
                defineProperty(textNode, 'children', {
                    value: Array.prototype.slice.call(node.shadowRoot.childNodes),
                });
                lightChildren.unshift(textNode);
            }
            return lightChildren;
        },
        configurable: true,
    });

    const result = DOMElement.serialize(
        node,
        config,
        indentation,
        depth,
        refs,
        (currentNode, currentConfig, currentIndentation, currentDepth, currentRefs) => {
            if (currentNode.textContent === '#shadow-root(open)') {
                return [
                    '#shadow-root(open)',
                    printChildren(
                        currentNode.children,
                        currentConfig,
                        currentIndentation + config.indent,
                        currentDepth + 1,
                        currentRefs,
                        printer
                    ),
                ].join('');
            }
            return printer(
                currentNode,
                currentConfig,
                currentIndentation,
                currentDepth,
                currentRefs
            );
        }
    );
    delete node.childNodes;

    return result;
}

module.exports.test = test;
module.exports.serialize = serialize;
