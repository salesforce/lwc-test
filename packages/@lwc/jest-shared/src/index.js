const knownScopedCssFiles = new Set();

/**
 * Indicate that this file is a scoped CSS file.
 * @param filename - full absolute file path
 */
function addKnownScopedCssFile(filename) {
    knownScopedCssFiles.add(filename);
}

/**
 * Check if this file is a scoped CSS file.
 * @param filename - full absolute file path
 * @returns {boolean} - true if scoped
 */
function isKnownScopedCssFile(filename) {
    return knownScopedCssFiles.has(filename);
}

module.exports = {
    addKnownScopedCssFile,
    isKnownScopedCssFile
};
