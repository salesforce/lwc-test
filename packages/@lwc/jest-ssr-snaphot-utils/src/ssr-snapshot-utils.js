const { renderComponent: lwcRenderComponent } = require('@lwc/engine-server');
const { createHash } = require('crypto');
const { readdirSync, readFileSync } = require('fs');
const { join, dirname, basename, extname } = require('path');
const { serverSideRenderComponent } = require('@lwc/ssr-runtime');

/**
 * Renders the component's markup, captures it in a snapshot that has a unique snapshot hash.
 *
 * @param {string} tagName - The name of the HTML tag or component.
 * @param {Function} Ctor - The constructor of the Lightning Web Component.
 * @param {Object} [props={}] - An object representing the properties of the component.
 * @param {Object} [customTestEnv={}] - An object representing the custom test env where the component is being validated.
 * @returns {{renderedComponent: string, snapshotHash: string}} - An object containing the rendered markup and the generated snapshot hash.
 */
async function renderAndHashComponent(tagName, Ctor, props = {}, customTestEnv = {}) {
    const snapshotHash = generateSnapshotHash(tagName, props, customTestEnv);
    const ssrMode = process.env.LWC_SSR_MODE || 'v2';

    const renderedComponent =
        ssrMode !== 'v1'
            ? await serverSideRenderComponent(tagName, Ctor, props)
            : lwcRenderComponent(tagName, Ctor, props);

    return { renderedComponent, snapshotHash };
}

/**
 * Reads the content of a snapshot file and returns the value associated with a specific hash.
 * Throws an error if the snapshot with the specified hash is not found.
 *
 * @param {string} tagName - The name of the HTML tag or component.
 * @param {Object} [props={}] - An object representing the properties of the component.
 * @param {Object} [customTestEnv={}] - An object representing the custom test env where the component is being validated.
 * @returns {string} - The markup associated with the given hash from the snapshot.
 * @throws {Error} - If the snapshot with the specified hash is not found.
 */
function readSnapshotMarkup(tagName, props = {}, customTestEnv = {}) {
    const testAbsPath = global.testFilePath;
    if (!testAbsPath) {
        throw new Error(
            'Test file path must be available in the global context. Make sure you utilize the custom JSDOM environment for SSR tests.'
        );
    }

    const snapshotPath = findFileByPrefix(testAbsPath);
    const fileContent = readFileSync(snapshotPath, 'utf8');
    const snapshotHash = generateSnapshotHash(tagName, props, customTestEnv);

    const regexPattern = `exports\\[\\\`[^\\\`]*${snapshotHash}[^\\\`]*\\\`] = \\\`([^\\\`]*)\\\`;`;
    const match = new RegExp(regexPattern).exec(fileContent);

    if (!match) {
        throw new Error(
            `Snapshot with hash ${snapshotHash} not found. Ensure the SSR tests have been run to generate the snapshot.`
        );
    }

    return match[1];
}

/**
 * Generates a SHA-256 hash based on the provided tagName, props, and custom test env.
 *
 * @param {string} tagName - The name of the HTML tag or component.
 * @param {Object} [props={}] - An object representing the properties of the component.
 * @param {Object} [customTestEnv={}] - An object representing the custom test env where the component is being validated.
 * @returns {string} - A hexadecimal string representing the generated hash.
 */
function generateSnapshotHash(tagName, props = {}, customTestEnv = {}) {
    return createHash('sha256')
        .update(tagName)
        .update(JSON.stringify(props))
        .update(JSON.stringify(customTestEnv))
        .digest('hex');
}

/**
 * Finds a snapshot file in the '__snapshots__' directory within the test suite directory
 * that starts with the test file name before the first period (including the period), for instance:
 * test file name: actionButton.server.test.ts, prefix: actionButton.
 *
 * @param {string} testSuiteAbsPath - The absolute path of the test suite file.
 * @returns {string} - The absolute path to the found snapshot file.
 * @throws {Error} - If no matching file is found or if an error occurs reading the directory.
 */
function findFileByPrefix(testSuiteAbsPath) {
    const testFileDir = dirname(testSuiteAbsPath);
    const snapshotsDir = join(testFileDir, '__snapshots__');

    // Extract the first portion of the filename including the period
    const baseName = basename(testSuiteAbsPath, extname(testSuiteAbsPath));
    const periodIndex = baseName.indexOf('.');
    const fileNamePrefix = periodIndex === -1 ? baseName : baseName.substring(0, periodIndex + 1);

    try {
        const matchedFile = readdirSync(snapshotsDir).find((file) =>
            file.startsWith(fileNamePrefix)
        );

        if (!matchedFile) {
            throw new Error(
                `File starting with prefix '${fileNamePrefix}' not found in directory '${snapshotsDir}'.`
            );
        }

        return join(snapshotsDir, matchedFile);
    } catch (error) {
        console.error(`Error reading directory: ${error.message}`);
        throw error;
    }
}

module.exports = { renderAndHashComponent, readSnapshotMarkup };
