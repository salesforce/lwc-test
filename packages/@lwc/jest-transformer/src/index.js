/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
const path = require('path');
const crypto = require('crypto');

const babelCore = require('@babel/core');
const lwcCompiler = require('@lwc/compiler');
const jestPreset = require('babel-preset-jest');
const babelCommonJs = require('@babel/plugin-transform-modules-commonjs');
const babelClassProperties = require('@babel/plugin-syntax-class-properties');
const babelDynamicImport = require('babel-plugin-transform-dynamic-import');
const babelSyntaxDecorators = require('@babel/plugin-syntax-decorators');
const babelTsPreset = require.resolve('@babel/preset-typescript');

const compilerVersion = require('@lwc/compiler/package.json').version;

const apexScopedImport = require('./transforms/apex-scoped-import');
const apexContinuationScopedImport = require('./transforms/apex-continuation-scoped-import');
const customPermissionImport = require('./transforms/custom-permission-scoped-import');
const i18nScopedImport = require('./transforms/i18n-scoped-import');
const labelScopedImport = require('./transforms/label-scoped-import');
const resourceScopedImport = require('./transforms/resource-scoped-import');
const contentAssetUrlScopedImport = require('./transforms/content-asset-url-scoped-import');
const schemaScopedImport = require('./transforms/schema-scoped-import');
const userScopedImport = require('./transforms/user-scoped-import');
const userPermissionImport = require('./transforms/user-permission-scoped-import');
const clientScopedImport = require('./transforms/client-scoped-import');
const messageChannelScopedImport = require('./transforms/message-channel-scoped-import');
const accessCheck = require('./transforms/access-check-scoped-import');

const BABEL_TS_CONFIG = {
    sourceMaps: 'inline',
    babelrc: false,
    configFile: false,
    plugins: [
        babelClassProperties,
        [
            babelSyntaxDecorators,
            {
                decoratorsBeforeExport: true,
            },
        ],
    ],
    presets: [babelTsPreset],
};

const BABEL_CONFIG = {
    sourceMaps: 'both',
    presets: [jestPreset],
    plugins: [
        babelCommonJs,
        babelDynamicImport,
        apexScopedImport,
        apexContinuationScopedImport,
        customPermissionImport,
        i18nScopedImport,
        labelScopedImport,
        contentAssetUrlScopedImport,
        resourceScopedImport,
        schemaScopedImport,
        userScopedImport,
        userPermissionImport,
        clientScopedImport,
        messageChannelScopedImport,
        accessCheck,
    ],
};

function isTypeScript(filePath) {
    return path.extname(filePath) === '.ts';
}

function transformTypeScript(src, filePath) {
    const { code } = babelCore.transform(src, {
        ...BABEL_TS_CONFIG,
        filename: filePath,
    });
    return code;
}

module.exports = {
    process(src, filePath) {
        if (isTypeScript(filePath)) {
            src = transformTypeScript(src, filePath);
        }

        // Set default module name and namespace value for the namespace because it can't be properly guessed from the path
        const { code, map } = lwcCompiler.transformSync(src, filePath, {
            name: 'test',
            namespace: 'x',
            outputConfig: {
                sourcemap: true,
            },
            experimentalDynamicComponent: {
                strictSpecifier: false,
            },
        });

        // if is not .js, we add the .compiled extension in the sourcemap
        const filename = path.extname(filePath) === '.js' ? filePath : filePath + '.compiled';
        // **Note: .html and .css don't return valid sourcemaps cause they are used for rollup
        const config = map && map.version ? { inputSourceMap: map } : {};

        return babelCore.transform(code, { ...BABEL_CONFIG, ...config, filename });
    },
    getCacheKey(fileData, filePath, configStr, options) {
        const { NODE_ENV } = process.env;
        return crypto
            .createHash('md5')
            .update(JSON.stringify(options), 'utf8')
            .update(fileData + filePath + configStr + NODE_ENV + compilerVersion, 'utf8')
            .digest('hex');
    },
};
