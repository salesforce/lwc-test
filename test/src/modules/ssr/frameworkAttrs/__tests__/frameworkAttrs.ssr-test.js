/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
import { renderAndHashComponent } from '@lwc/jest-ssr-snapshot-utils/src/ssr-snapshot-utils';
import FrameworkAttrs from '../frameworkAttrs';

it('serializes component with framework-supplied attributes', () => {
    const { renderedComponent, snapshotHash } = renderAndHashComponent(
        'x-basic',
        FrameworkAttrs,
        {}
    );

    expect(renderedComponent).toMatchSnapshot(snapshotHash);
});
