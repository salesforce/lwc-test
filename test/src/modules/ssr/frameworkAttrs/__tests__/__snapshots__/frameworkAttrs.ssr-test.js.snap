// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serializes component with framework-supplied attributes 1`] = `
<x-basic>
  <style type="text/css">
    h1 {color: blue;}
  </style>
  <h1>
    Hello world
  </h1>
</x-basic>
`;
